src/stan/mcmc/hmc/hamiltonians/base_hamiltonian.hpp

```diff
@@ -37,29 +37,50 @@ namespace stan {
       }
 
       // The time derivative of the virial, G = \sum_{d = 1}^{D} q^{d} p_{d}.
-      virtual double dG_dt(Point& z) = 0;
+      virtual double dG_dt(
+        Point& z,
+        interface_callbacks::writer::base_writer& info_writer,
+        interface_callbacks::writer::base_writer& error_writer) = 0;
 
       // tau = 0.5 p_{i} p_{j} Lambda^{ij} (q)
-      virtual const Eigen::VectorXd dtau_dq(Point& z) = 0;
+      virtual const Eigen::VectorXd dtau_dq(
+        Point& z,
+        interface_callbacks::writer::base_writer& info_writer,
+        interface_callbacks::writer::base_writer& error_writer) = 0;
 
       virtual const Eigen::VectorXd dtau_dp(Point& z) = 0;
 
       // phi = 0.5 * log | Lambda (q) | + V(q)
-      virtual const Eigen::VectorXd dphi_dq(Point& z) = 0;
+      virtual const Eigen::VectorXd dphi_dq(
+        Point& z,
+        interface_callbacks::writer::base_writer& info_writer,
+        interface_callbacks::writer::base_writer& error_writer) = 0;
 
       virtual void sample_p(Point& z, BaseRNG& rng) = 0;
 
-      virtual void
-      init(Point& z,
-           interface_callbacks::writer::base_writer& info_writer,
-           interface_callbacks::writer::base_writer& error_writer) {
-        this->update(z, info_writer, error_writer);
+      void init(Point& z,
+                interface_callbacks::writer::base_writer& info_writer,
+                interface_callbacks::writer::base_writer& error_writer) {
+        this->update_potential_gradient(z, info_writer, error_writer);
       }
 
-      virtual void
-      update(Point& z,
-             interface_callbacks::writer::base_writer& info_writer,
-             interface_callbacks::writer::base_writer& error_writer) {
+      void update_potential(
+        Point& z,
+        interface_callbacks::writer::base_writer& info_writer,
+        interface_callbacks::writer::base_writer& error_writer) {
+        try {
```

Comment:
(fix now) Negation is faster (much much faster) and I think it's easier to read.